src/envoy/http/authn/http_filter.cc

```diff
@@ -30,59 +57,207 @@ void AuthenticationFilter::onDestroy() {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
 }
 
-FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
+                                                        bool) {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
+  state_ = IstioAuthN::State::PROCESSING;
+  if (config_.peers_size() == 0) {
+    ENVOY_LOG(debug, "No method defined. Skip source authentication.");
+    onAuthenticatePeerDone(&headers, 0, nullptr, true);
+  } else {
+    authenticatePeer(
+        headers, config_.peers(0),
+        std::bind(&AuthenticationFilter::onAuthenticatePeerDone, this, &headers,
+                  0, std::placeholders::_1, std::placeholders::_2));
+  }
+  if (state_ == IstioAuthN::State::COMPLETE) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  stopped_ = true;
+  return FilterHeadersStatus::StopIteration;
+}
+
+void AuthenticationFilter::authenticatePeer(
+    HeaderMap& headers, const iaapi::PeerAuthenticationMethod& method,
+    const AuthenticateDoneCallback& done_callback) {
```

Comment:
don't understand why done_callback has to be passed in as function argument.

It seems that all places are using onAuthenticatePeerDone().    

Can this function directly call onAuthenticatePeerDone()?

Reply:
Because I want to bind the first few parameters of the callback to keeping state of the iteration.

Alternatively, we can have the class (AuthenticationFilter) having those variables as member variables, but personally I think it less clean as any methods can access/modify those variables. Having them pass as argument make sure only callback will use them.